components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -14,9 +18,78 @@ pub enum CanonicalizationResult {
 }
 
 pub struct LocaleCanonicalizer<'a> {
+    aliases: DataPayload<'a, AliasesV1>,
     likely_subtags: DataPayload<'a, LikelySubtagsV1>,
 }
 
+#[inline]
+fn uts35_rule_matches(source: &Locale, ruletype: &LanguageIdentifier) -> bool {
+    (ruletype.language.is_empty() || ruletype.language == source.id.language)
+        && (ruletype.script.is_none() || ruletype.script == source.id.script)
+        && (ruletype.region.is_none() || ruletype.region == source.id.region)
+        && (ruletype.variants.is_empty()
+            || ruletype
+                .variants
+                .iter()
+                .all(|v| source.id.variants.contains(v)))
+}
+
+fn uts35_replacement(
+    source: &mut Locale,
+    ruletype_has_language: bool,
+    ruletype_has_script: bool,
+    ruletype_has_region: bool,
+    ruletype_variants: Option<&subtags::Variants>,
+    replacement: &LanguageIdentifier,
+) {
+    if ruletype_has_language || source.id.language.is_empty() && !replacement.language.is_empty() {
+        source.id.language = replacement.language;
+    }
+    if ruletype_has_script || source.id.script.is_none() && replacement.script.is_some() {
+        source.id.script = replacement.script;
+    }
+    if ruletype_has_region || source.id.region.is_none() && replacement.region.is_some() {
+        source.id.region = replacement.region;
+    }
+    if let Some(ruletype_variants) = ruletype_variants {
+        let mut variants: Vec<subtags::Variant> = source
+            .id
+            .variants
+            .iter()
+            .filter(|x| !ruletype_variants.contains(x))
+            .copied()
+            .collect();
+        for variant in replacement.variants.iter() {
+            variants.push(*variant);
+        }
```

Comment:
`suggestion (non-blocking)`: this could be potentially sped up if we were to allocate optimistically `variants` for sum of lengths and insert using `binary_search` already pre-sorted. Not sure if it's worth it.

Reply:
My feeling is that it is probably not worth adding the complexity unless we start seeing very large numbers of variants.